### PR TITLE
ENH: SPMD interface for IncrementalEmpiricalCovariance

### PR DESCRIPTION
**Pull request by olegkkruglov**
### Description 

* Added SPMD interface for IncrementalEmpiricalCovariance
* Added example of its usage.
* Changed policy saving workflow, now queue is saved to attributes instead of policy. It is necessary because `finalize_fit` requires `spmd_policy`, but `partial_fit` requires `data_parallel_policy` on oneDAL side
* `finalize_fit` now uses provided queue for computations on onedal4py side. 
* Contains some content from https://github.com/intel/scikit-learn-intelex/pull/1777 for test implementation

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes, if necessary.
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

File: examples/sklearnex/incremental_covariance_spmd.py
Comment (olegkkruglov) [reply examples/sklearnex/incremental_covariance_spmd.py:28]: if I understood you correctly, this multiplication is applied in the next row

File: onedal/covariance/covariance.cpp
Comment (olegkkruglov) [reply onedal/covariance/covariance.cpp:181]: no, we don't have this instantiation on oneDAL side. communication between devices exists only in finalize.

File: sklearnex/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/covariance/incremental_covariance.py:290]: fixed

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:62]: BaseEmpiricalCovariance is inherited from BaseEstimator. The trick here is that we inherit from BaseEstimatorSPMD which also has _get_backend. Thus, if we put self._get_backend here then SPMD backend would be called which does not contain partial_compute_result.

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:90]: It is for finalize_fit dispatching. It does not have data argument, so, in case if user does not provide queue explicitly then the last queue from partial_fit is used.

File: examples/sklearnex/incremental_covariance_spmd.py
Comment (ethanglaser) [reply examples/sklearnex/incremental_covariance_spmd.py:28]: `(a * b) // c` does not necessarily equal `a * (b // c)`

Comment (olegkkruglov) [reply examples/sklearnex/incremental_covariance_spmd.py:28]: ok, but what is wrong with the example? `local_size` represents size of data on each rank, it is ceiling of integer division of number of rows by number of ranks. in the next row corresponding slice of size `local_size` is taken.

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/covariance/incremental_covariance.py:62]: I'd prefer a definition of _reset in the SPMD interface which uses a super call for locality code. That would be simpler for maintainers in the future to see why certain things are done.  At a minimum it needs a comment in the code.

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/covariance/incremental_covariance.py:90]: Can't the last queue be extracted from the stored policy via the `_queue` property? A lot of this logic might be unnecessary. @olegkkruglov let me know.

**Review by icfaust — COMMENTED** [review]
Some high level questions about the policy queues and the backend.  Also comments would be nice.

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:90]: It was done like that before this PR. but now it turned out that different policies must be used in finalize_fit and partial_fit that's why stored policy from partial fit is not acceptable

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:62]: It is also necessary to redefine `partial_fit` because `partial_compute` also does not exist in the SPMD backend. I'm not sure if the code duplication is better idea than the currently implemented one

Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:62]: and also have to add that super call would not work there because BaseEstimator and BaseEstimatorSPMD have the same methods and if the class has both of them as parents we anyway need to specify directly which of them should be called.

File: onedal/_device_offload.py
Comment (olegkkruglov) [reply onedal/_device_offload.py]: unfortunately this one is already merged.

File: onedal/_device_offload.py
Comment (samir-nasibli) [reply onedal/_device_offload.py]: Will refactor it on #1861

**Comment by olegkkruglov** [comment]
https://ecmd.jf.intel.com/commander/link/jobDetails/jobs/ef4423ff-9ba8-f1d0-abf8-a4bf010d0e2e

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [reply onedal/covariance/incremental_covariance.py:62]: I would ask you do refactoring after this PR merged. Currently this looks like a workaround.
Create some `BaseIncreamenatlEstimator`, where get_backend method depending on provided method name returns required backend. This will be common for all incremental algos.

Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:62]: > Create some `BaseIncreamenatlEstimator`, where get_backend method depending on provided method name returns required backend. This will be common for all incremental algos.

this might be a good idea, I'll think about it

File: onedal/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/covariance/incremental_covariance.py:90]: Since it's just partial_fit and reset which requires a data parallel policy, I would overload partial_fit and reset in the spmd interface, and make sure that the correct policy is taken there via a super call.  A small duplication of code, but at least its clear to the developer and the user on what is going on, and its located in a place that someone looking to understand the spmd interface can see the limitations of partial_fit for spmd in the incremental algos. It would be then more straightforward for @samir-nasibli 's request for a refactor with an Incremental base class.

**Review by icfaust — COMMENTED** [review]
I think the overloading partial_fit and _reset in the SPMD class is more preferrable to changes in underlying classes and storing queues. It would be simpler to understand, spmd problems should stay to spmd.

**Comment by olegkkruglov** [comment]
> I think the overloading partial_fit and _reset in the SPMD class is more preferrable to changes in underlying classes and storing queues. It would be simpler to understand, spmd problems should stay to spmd.

Why is storing queue worse than storing policy? As far as I see, storing queue instead of policy is unavoidable because the policy from partial_fit can't be used in finalize_fit and it is the only thing the policy was stored for.

File: examples/sklearnex/incremental_covariance_spmd.py
Comment (ethanglaser) [reply examples/sklearnex/incremental_covariance_spmd.py:28]: my mistake - I just wrote out an example and this makes sense.

File: examples/sklearnex/incremental_covariance_spmd.py
Comment (ethanglaser) [inline examples/sklearnex/incremental_covariance_spmd.py:49]: will finalize_fit be called? since this is the first example of incremental spmd some comments would be good to explain how it works, even I am a bit confused

File: sklearnex/spmd/covariance/incremental_covariance.py
Comment (ethanglaser) [inline sklearnex/spmd/covariance/incremental_covariance.py:27]: What does this do?

File: sklearnex/covariance/incremental_covariance.py
Comment (ethanglaser) [reply sklearnex/covariance/incremental_covariance.py:290]: I guess more appropriate question is - does it need to be passed to this function? it looks like it does not get used

File: sklearnex/covariance/incremental_covariance.py
Comment (ethanglaser) [inline sklearnex/covariance/incremental_covariance.py:116]: How is queue obtained when called from line 136 or 147?

File: onedal/spmd/covariance/incremental_covariance.py
Comment (ethanglaser) [inline onedal/spmd/covariance/incremental_covariance.py:36]: how does this differ than the _reset that is inherited from batch IncrementalCovariance?

File: onedal/spmd/covariance/incremental_covariance.py
Comment (ethanglaser) [inline onedal/spmd/covariance/incremental_covariance.py:40]: can super().partial_fit(...) be used instead?

File: sklearnex/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/spmd/covariance/incremental_covariance.py:27]: it changes _onedal_incremental_covariance to spmd version

File: onedal/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/spmd/covariance/incremental_covariance.py:36]: Inherited reset would call spmd _get_backend from BaseEstimatorSPMD

File: onedal/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/spmd/covariance/incremental_covariance.py:40]: the same reason as here
https://github.com/intel/scikit-learn-intelex/pull/1941#discussion_r1687701688

File: sklearnex/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/covariance/incremental_covariance.py:116]: it takes queue stored from partial_fit call in attributes of onedal estimator

File: onedal/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/spmd/covariance/incremental_covariance.py:40]: I thought I fulfilled what you asked for here https://github.com/intel/scikit-learn-intelex/pull/1941#discussion_r1682455687
Let's discuss it today as well, maybe I understood your request wrongly.

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [reply onedal/covariance/incremental_covariance.py:89]: > It is for finalize_fit dispatching. It does not have data argument, so, in case if user does not provide queue explicitly then the last queue from partial_fit is used.

Maybe it make sense explicitly ask user provide sycl queue? Otherwise it is headache

Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:89]: finalize_fit is called implicitly on sklearnex side. if we want to keep scikit-like interface (without explicit finalize call) the only option is to call finalize after every call of partial_fit. this option was rejected on arch meeting, that's why keeping queue in attributes is currently unavoidable.

File: onedal/covariance/incremental_covariance.py
Comment (ethanglaser) [inline onedal/covariance/incremental_covariance.py:130]: Logic in finalize_fit and partial_fit does not match - can we align this? In terms of getting policy from queue/self._queue

File: sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py
Comment (olegkkruglov) [reply sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py:89]: fit should be tested anyway, because it is different interface which can potentially contain errors. I can't get how the second option reducing verbosity. the number of tests would remain the same. moreover, and why do you want to reduce test verbosity?

File: onedal/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/spmd/covariance/incremental_covariance.py:61]: what do you mean by occurs in super? it is not called in any super call across this method.

File: sklearnex/spmd/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply sklearnex/spmd/covariance/incremental_covariance.py:29]: this attribute is used on sklearnex side to call oneDAL via onedal4py estimator. here regular onedal4py estimator replaced by onedal4py SPMD estimator. it allows to use the same API as for regular sklearnex estimator but SPMD oneDAL part is called on backend.

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:130]: it shouldn't match. in partial_fit queue is supposed to be provided explicitly from sklearnex side via data. finalize_fit is designed to handle the case queue is not provided. it takes it from self._queue in this case assuming the same device as for partial_fit calls must be used .

**Comment by ethanglaser** [comment]
Please resolve fails in internal CI - specifically it looks like assume_centered=True tests fail at means check

File: onedal/spmd/covariance/incremental_covariance.py
Comment (icfaust) [inline onedal/spmd/covariance/incremental_covariance.py:23]: https://github.com/intel/scikit-learn-intelex/pull/1961#discussion_r1732104503

File: onedal/spmd/covariance/incremental_covariance.py
Comment (icfaust) [inline onedal/spmd/covariance/incremental_covariance.py:89]: https://github.com/intel/scikit-learn-intelex/pull/1961#discussion_r1732107497

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (icfaust) [inline sklearnex/covariance/tests/test_incremental_covariance.py:44]: doesn't there exist an is_gpu attribute on a sycl_device (https://github.com/intel/scikit-learn-intelex/blob/main/onedal/_device_offload.py#L50)? Should allow you to remove the try except and remove the codefactor issue with the ```except```

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (icfaust) [inline sklearnex/covariance/tests/test_incremental_covariance.py:48]: I'd leave something a bit more informative here. What kind of bug? incorrect results, memory corruption, seg fault?

File: sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py
Comment (icfaust) [inline sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py:61]: https://github.com/intel/scikit-learn-intelex/pull/1961#discussion_r1732087952

If this is being used as a default 'gold_data' for all testing for more estimators you may want to move it into _utils_spmd.py

**Review by icfaust — COMMENTED** [review]
Lots of referencing to this review: https://github.com/intel/scikit-learn-intelex/pull/1961#pullrequestreview-2262044792  I would say all the SPMD interface PRs (#1979 #1972 #1961 and #1941) should be analyzed and changed to the same standard (i.e. comments here and on #1961 will likely apply to all).

File: onedal/spmd/covariance/incremental_covariance.py
Comment (icfaust) [reply onedal/spmd/covariance/incremental_covariance.py:61]: I just had to spend some minutes verifying this.  As an aside, we really got to set a standard of where the finiteness check occurs. I am doing this check at least 3 times a week.

File: sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py
Comment (olegkkruglov) [reply sklearnex/spmd/covariance/tests/test_incremental_covariance_spmd.py:61]: https://github.com/intel/scikit-learn-intelex/pull/1961#discussion_r1732638526

**Review by bdmoore1 — APPROVED** [review]

File: onedal/covariance/incremental_covariance.py
Comment (samir-nasibli) [inline onedal/covariance/incremental_covariance.py:129]: I think at finalize_fit call self._queue should be deleted.

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (samir-nasibli) [inline sklearnex/covariance/tests/test_incremental_covariance.py:155]:
```suggestion
            "Due to a bug on oneDAL side, means are  not set to zero when assume_centered=True"
```

**Review by samir-nasibli — COMMENTED** [review]
Please share latest internal CI run

File: onedal/covariance/incremental_covariance.py
Comment (olegkkruglov) [reply onedal/covariance/incremental_covariance.py:128]: https://github.com/intel/scikit-learn-intelex/pull/1979#discussion_r1738454650

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (icfaust) [inline sklearnex/covariance/tests/test_incremental_covariance.py:42]:
```suggestion
    if assume_centered and is_gpu and not daal_check_version((2025, "P", 0)):
```

File: sklearnex/covariance/tests/test_incremental_covariance.py
Comment (icfaust) [inline sklearnex/covariance/tests/test_incremental_covariance.py:153]:
```suggestion
    if assume_centered and is_gpu and not daal_check_version((2025, "P", 0)):
```

**Review by icfaust — COMMENTED** [review]
Generally good to go,  given the issues observed with dtypes in #1961, could you also parametrize dtype here just to see what it does to the results? (No dtype assert necessary this time)

**Comment by olegkkruglov** [comment]
> Generally good to go, given the issues observed with dtypes in #1961, could you also parametrize dtype here just to see what it does to the results? (No dtype assert necessary this time)

done

File: examples/sklearnex/incremental_covariance_spmd.py
Comment (samir-nasibli) [inline examples/sklearnex/incremental_covariance_spmd.py:54]: Partial fit is using only non-spmd policy, how it then offloads each of GPU?

Comment (samir-nasibli) [reply examples/sklearnex/incremental_covariance_spmd.py:54]: I can answer my question. When we run mpirun -n 2 for example, then we are creating queue for each of the device, the calling partial_fit for separated data by the rank. and we have partial results on the each of the device.
for `cov.partial_fit(dpt_X)` we are using dataparallel policy inside, but for `cov.covariance_` we are using spmd_policy, since the communication is required between devices when we calling final fit.

**Review by icfaust — APPROVED** [review]
Approval dependent on @samir-nasibli 's requests (docstrings at least).

**Review by samir-nasibli — APPROVED** [review]
Thank you @olegkkruglov !
Now looks good to me!
Assuming green CI, please share internal CI job link.

Expecting quick follow up refactoring, based on the tickets created and for the docstrings mentioned.
I am good to go functionally with this PR as is, but the refactoring is deserved before CF.

**Pull request merged**
